File: macosx/GroupsPrefsController.m
/******************************************************************************
 * $Id$
 *
 * Copyright (c) 2007-2012 Transmission authors and contributors
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *****************************************************************************/

#import "GroupsPrefsController.h"
#import "GroupsController.h"
#import "ExpandedPathToPathTransformer.h"
#import "ExpandedPathToIconTransformer.h"
#import "NSApplicationAdditions.h"

#define GROUP_TABLE_VIEW_DATA_TYPE @"GroupTableViewDataType"

#define ADD_TAG 0
#define REMOVE_TAG 1

@interface GroupsPrefsController (Private)

- (void) updateSelectedGroup;
- (void) refreshCustomLocationWithSingleGroup;

@end

@implementation GroupsPrefsController

- (void) awakeFromNib
{
    [fTableView registerForDraggedTypes: [NSArray arrayWithObject: GROUP_TABLE_VIEW_DATA_TYPE]];
    
    [fSelectedColorView addObserver: self forKeyPath: @"color" options: 0 context: NULL];
    
    [self updateSelectedGroup];
}

- (NSInteger) numberOfRowsInTableView: (NSTableView *) tableview
{
    return [[GroupsController groups] numberOfGroups];
}

- (id) tableView: (NSTableView *) tableView objectValueForTableColumn: (NSTableColumn *) tableColumn row: (NSInteger) row
{
    GroupsController * groupsController = [GroupsController groups];
    NSInteger groupsIndex = [groupsController indexForRow: row];
    
    NSString * identifier = [tableColumn identifier];
    if ([identifier isEqualToString: @"Color"])
        return [groupsController imageForIndex: groupsIndex];
    else
        return [groupsController nameForIndex: groupsIndex];
}

- (void) tableViewSelectionDidChange: (NSNotification *) notification
{
    [self updateSelectedGroup];
}

- (void) observeValueForKeyPath: (NSString *) keyPath ofObject: (id) object change: (NSDictionary *) change context: (void *) context
{
    if (object == fSelectedColorView && [fTableView numberOfSelectedRows] == 1)
    {
       NSInteger index = [[GroupsController groups] indexForRow: [fTableView selectedRow]];
       [[GroupsController groups] setColor: [fSelectedColorView color] forIndex: index];
       [fTableView setNeedsDisplay: YES];
    }
}

- (void) controlTextDidEndEditing: (NSNotification *) notification
{
    if ([notification object] == fSelectedColorNameField)
    {
       NSInteger index = [[GroupsController groups] indexForRow: [fTableView selectedRow]];
       [[GroupsController groups] setName: [fSelectedColorNameField stringValue] forIndex: index];
       [fTableView setNeedsDisplay: YES];
    }
}

- (BOOL) tableView: (NSTableView *) tableView writeRowsWithIndexes: (NSIndexSet *) rowIndexes toPasteboard: (NSPasteboard *) pboard
{
    [pboard declareTypes: [NSArray arrayWithObject: GROUP_TABLE_VIEW_DATA_TYPE] owner: self];
    [pboard setData: [NSKeyedArchiver archivedDataWithRootObject: rowIndexes] forType: GROUP_TABLE_VIEW_DATA_TYPE];
    return YES;
}

- (NSDragOperation) tableView: (NSTableView *) tableView validateDrop: (id <NSDraggingInfo>) info
    proposedRow: (NSInteger) row proposedDropOperation: (NSTableViewDropOperation) operation
{
    NSPasteboard * pasteboard = [info draggingPasteboard];
    if ([[pasteboard types] containsObject: GROUP_TABLE_VIEW_DATA_TYPE])
    {
        [fTableView setDropRow: row dropOperation: NSTableViewDropAbove];
        return NSDragOperationGeneric;
    }
    
    return NSDragOperationNone;
}

- (BOOL) tableView: (NSTableView *) tableView acceptDrop: (id <NSDraggingInfo>) info row: (NSInteger) newRow
    dropOperation: (NSTableViewDropOperation) operation
{
    NSPasteboard * pasteboard = [info draggingPasteboard];
    if ([[pasteboard types] containsObject: GROUP_TABLE_VIEW_DATA_TYPE])
    {
        NSIndexSet * indexes = [NSKeyedUnarchiver unarchiveObjectWithData: [pasteboard dataForType: GROUP_TABLE_VIEW_DATA_TYPE]];
        NSInteger oldRow = [indexes firstIndex];
        
        if (oldRow < newRow)
            newRow--;

        [fTableView beginUpdates];
        
        [[GroupsController groups] moveGroupAtRow: oldRow toRow: newRow];

        [fTableView moveRowAtIndex: oldRow toIndex: newRow];
        [fTableView endUpdates];
    }
    
    return YES;
}

- (void) addRemoveGroup: (id) sender
{
    if ([NSColorPanel sharedColorPanelExists])
        [[NSColorPanel sharedColorPanel] close];

    NSInteger row;
    
    switch ([[sender cell] tagForSegment: [sender selectedSegment]])
    {
        case ADD_TAG:
            [fTableView beginUpdates];
            
            [[GroupsController groups] addNewGroup];
            
            row = [fTableView numberOfRows];

            [fTableView insertRowsAtIndexes: [NSIndexSet indexSetWithIndex: row] withAnimation: NSTableViewAnimationSlideUp];
            [fTableView endUpdates];
            
            [fTableView selectRowIndexes: [NSIndexSet indexSetWithIndex: row] byExtendingSelection: NO];
            [fTableView scrollRowToVisible: row];
            
            [[fSelectedColorNameField window] makeFirstResponder: fSelectedColorNameField];
            
            break;
        
        case REMOVE_TAG:
            row = [fTableView selectedRow];
            

            [fTableView beginUpdates];
            
            [[GroupsController groups] removeGroupWithRowIndex: row];            

            [fTableView removeRowsAtIndexes: [NSIndexSet indexSetWithIndex: row] withAnimation: NSTableViewAnimationSlideUp];
            [fTableView endUpdates];
            
            if ([fTableView numberOfRows] > 0)
            {
                if (row == [fTableView numberOfRows])
                    --row;
                [fTableView selectRowIndexes: [NSIndexSet indexSetWithIndex: row] byExtendingSelection: NO];
                [fTableView scrollRowToVisible: row];
            }
            
            break;
    }
    
    [self updateSelectedGroup];
}

- (void) customDownloadLocationSheetShow: (id) sender
{
    NSOpenPanel * panel = [NSOpenPanel openPanel];

    [panel setPrompt: NSLocalizedString(@"Select", "Preferences -> Open panel prompt")];
    [panel setAllowsMultipleSelection: NO];
    [panel setCanChooseFiles: NO];
    [panel setCanChooseDirectories: YES];
    [panel setCanCreateDirectories: YES];
    
    [panel beginSheetModalForWindow: [fCustomLocationPopUp window] completionHandler: ^(NSInteger result) {
        const NSInteger index = [[GroupsController groups] indexForRow: [fTableView selectedRow]];
        if (result == NSFileHandlingPanelOKButton)
        {
            NSString * path = [[[panel URLs] objectAtIndex: 0] path];
            [[GroupsController groups] setCustomDownloadLocation: path forIndex: index];
            [[GroupsController groups] setUsesCustomDownloadLocation: YES forIndex: index];
        }
        else
        {
            if (![[GroupsController groups] customDownloadLocationForIndex: index])
                [[GroupsController groups] setUsesCustomDownloadLocation: NO forIndex: index];
        }
        
        [self refreshCustomLocationWithSingleGroup];
        
        [fCustomLocationPopUp selectItemAtIndex: 0];
    }];
}

- (IBAction) toggleUseCustomDownloadLocation: (id) sender
{
    NSInteger index = [[GroupsController groups] indexForRow: [fTableView selectedRow]];
    if ([fCustomLocationEnableCheck state] == NSOnState)
    {
        if ([[GroupsController groups] customDownloadLocationForIndex: index])
            [[GroupsController groups] setUsesCustomDownloadLocation: YES forIndex: index];
        else
            [self customDownloadLocationSheetShow: nil];
    }
    else
        [[GroupsController groups] setUsesCustomDownloadLocation: NO forIndex: index];

    [fCustomLocationPopUp setEnabled: ([fCustomLocationEnableCheck state] == NSOnState)];
}

#pragma mark -
#pragma mark Rule editor

- (IBAction) toggleUseAutoAssignRules: (id) sender;
{
    NSInteger index = [[GroupsController groups] indexForRow: [fTableView selectedRow]];
    if ([fAutoAssignRulesEnableCheck state] == NSOnState)
    {
        if ([[GroupsController groups] autoAssignRulesForIndex: index])
            [[GroupsController groups] setUsesAutoAssignRules: YES forIndex: index];
        else
            [self orderFrontRulesSheet: nil];
    }
    else
        [[GroupsController groups] setUsesAutoAssignRules: NO forIndex: index];

    [fAutoAssignRulesEditButton setEnabled: [fAutoAssignRulesEnableCheck state] == NSOnState];
}

- (IBAction) orderFrontRulesSheet: (id) sender;
{
    if (!fGroupRulesSheetWindow)
        [NSBundle loadNibNamed: @"GroupRules" owner: self];

    NSInteger index = [[GroupsController groups] indexForRow: [fTableView selectedRow]];
	NSPredicate *predicate = [[GroupsController groups] autoAssignRulesForIndex: index];
	[fRuleEditor setObjectValue: predicate];
	
    if ([fRuleEditor numberOfRows] == 0)
        [fRuleEditor addRow: nil];
        
    [NSApp beginSheet: fGroupRulesSheetWindow modalForWindow: [fTableView window] modalDelegate: nil didEndSelector: NULL
        contextInfo: NULL];
}

- (IBAction) cancelRules: (id) sender;
{
    [fGroupRulesSheetWindow orderOut: nil];
    [NSApp endSheet: fGroupRulesSheetWindow];
    
    NSInteger index = [[GroupsController groups] indexForRow: [fTableView selectedRow]];
    if (![[GroupsController groups] autoAssignRulesForIndex: index])
    {
        [[GroupsController groups] setUsesAutoAssignRules: NO forIndex: index];
        [fAutoAssignRulesEnableCheck setState: NO];
        [fAutoAssignRulesEditButton setEnabled: NO];
    }
}

- (IBAction) saveRules: (id) sender;
{
    [fGroupRulesSheetWindow orderOut: nil];
    [NSApp endSheet: fGroupRulesSheetWindow];
    
    NSInteger index = [[GroupsController groups] indexForRow: [fTableView selectedRow]];
    [[GroupsController groups] setUsesAutoAssignRules: YES forIndex: index];
    
    NSPredicate * predicate = [fRuleEditor objectValue];
    [[GroupsController groups] setAutoAssignRules: predicate forIndex: index];
	
    [fAutoAssignRulesEnableCheck setState: [[GroupsController groups] usesAutoAssignRulesForIndex: index]];
    [fAutoAssignRulesEditButton setEnabled: [fAutoAssignRulesEnableCheck state] == NSOnState];
}

- (void) ruleEditorRowsDidChange: (NSNotification *) notification
{
    NSScrollView * ruleEditorScrollView = [fRuleEditor enclosingScrollView];
    
    const CGFloat rowHeight = [fRuleEditor rowHeight];
    const CGFloat bordersHeight = [ruleEditorScrollView frame].size.height - [ruleEditorScrollView contentSize].height;

    const CGFloat requiredRowCount = [fRuleEditor numberOfRows];
    const CGFloat maxVisibleRowCount = (long)((NSHeight([[[fRuleEditor window] screen] visibleFrame]) * 2 / 3) / rowHeight);
    
    [fRuleEditorHeightConstraint setConstant: MIN(requiredRowCount, maxVisibleRowCount) * rowHeight + bordersHeight];
    [ruleEditorScrollView setHasVerticalScroller: requiredRowCount > maxVisibleRowCount];
}

@end

@implementation GroupsPrefsController (Private)

- (void) updateSelectedGroup
{
    [fAddRemoveControl setEnabled: [fTableView numberOfSelectedRows] > 0 forSegment: REMOVE_TAG];
    if ([fTableView numberOfSelectedRows] == 1)
    {
        const NSInteger index = [[GroupsController groups] indexForRow: [fTableView selectedRow]];
        [fSelectedColorView setColor: [[GroupsController groups] colorForIndex: index]];
        [fSelectedColorView setEnabled: YES];
        [fSelectedColorNameField setStringValue: [[GroupsController groups] nameForIndex: index]];
        [fSelectedColorNameField setEnabled: YES];
        
        [self refreshCustomLocationWithSingleGroup];

        [fAutoAssignRulesEnableCheck setState: [[GroupsController groups] usesAutoAssignRulesForIndex: index]];
        [fAutoAssignRulesEnableCheck setEnabled: YES];
        [fAutoAssignRulesEditButton setEnabled: ([fAutoAssignRulesEnableCheck state] == NSOnState)];
    }
    else
    {
        [fSelectedColorView setColor: [NSColor whiteColor]];
        [fSelectedColorView setEnabled: NO];
        [fSelectedColorNameField setStringValue: @""];
        [fSelectedColorNameField setEnabled: NO];
        [fCustomLocationEnableCheck setEnabled: NO];
        [fCustomLocationPopUp setEnabled: NO];
        [fAutoAssignRulesEnableCheck setEnabled: NO];
        [fAutoAssignRulesEditButton setEnabled: NO];
    }
}

- (void) refreshCustomLocationWithSingleGroup
{
    const NSInteger index = [[GroupsController groups] indexForRow: [fTableView selectedRow]];
    
    const BOOL hasCustomLocation = [[GroupsController groups] usesCustomDownloadLocationForIndex: index];
    [fCustomLocationEnableCheck setState: hasCustomLocation];
    [fCustomLocationEnableCheck setEnabled: YES];
    [fCustomLocationPopUp setEnabled: hasCustomLocation];
    
    NSString * location = [[GroupsController groups] customDownloadLocationForIndex: index];
    if (location)
    {
        ExpandedPathToPathTransformer * pathTransformer = [[[ExpandedPathToPathTransformer alloc] init] autorelease];
        [[fCustomLocationPopUp itemAtIndex: 0] setTitle: [pathTransformer transformedValue: location]];
        ExpandedPathToIconTransformer * iconTransformer = [[[ExpandedPathToIconTransformer alloc] init] autorelease];
        [[fCustomLocationPopUp itemAtIndex: 0] setImage: [iconTransformer transformedValue: location]];
    }
    else
    {
        [[fCustomLocationPopUp itemAtIndex: 0] setTitle: @""];
        [[fCustomLocationPopUp itemAtIndex: 0] setImage: nil];
    }
}

@end

File: macosx/GroupsPrefsController.swift
// This file Copyright © 2007-2023 Transmission authors and contributors.
// It may be used under the MIT (SPDX: MIT) license.
// License text can be found in the licenses/ folder.

import AppKit

enum SegmentTag: Int {
    case add = 0
    case remove = 1
}

class GroupsPrefsController: NSObject {
    private static let kGroupTableViewDataType = NSPasteboard.PasteboardType("GroupTableViewDataType")

    @IBOutlet private weak var fTableView: NSTableView!
    @IBOutlet private weak var fAddRemoveControl: NSSegmentedControl!
    @IBOutlet private weak var fSelectedColorView: NSColorWell!
    @IBOutlet private weak var fSelectedColorNameField: NSTextField!
    @IBOutlet private weak var fCustomLocationEnableCheck: NSButton!
    @IBOutlet private weak var fCustomLocationPopUp: NSPopUpButton!
    @IBOutlet private weak var fAutoAssignRulesEnableCheck: NSButton!
    @IBOutlet private weak var fAutoAssignRulesEditButton: NSButton!
    @IBOutlet private weak var groupRulesSheetWindow: NSWindow!
    @IBOutlet private weak var ruleEditor: NSPredicateEditor!
    @IBOutlet private weak var ruleEditorHeightConstraint: NSLayoutConstraint!

    private var selectedColorViewObservation: NSKeyValueObservation?

    override func awakeFromNib() {
        super.awakeFromNib()
        fTableView.registerForDraggedTypes([ Self.kGroupTableViewDataType ])

        selectedColorViewObservation = observeSelectedColorView()

#if compiler(>=5.7.1)// macOS 13.0 support based on https://xcodereleases.com
        if #available(macOS 13.0, *) {
            fSelectedColorView.colorWellStyle = .minimal
        }
#endif

        updateSelectedGroup()
    }

    private func observeSelectedColorView() -> NSKeyValueObservation {
        return fSelectedColorView.observe(\NSColorWell.color) { [weak self] selectedColorView, _/*observedChange*/ in
            guard let self = self else {
                return
            }
            if self.fTableView.numberOfSelectedRows == 1 {
                let index = GroupsController.groups.index(forRow: self.fTableView.selectedRow)
                GroupsController.groups.setColor(selectedColorView.color, for: index)
                self.fTableView.needsDisplay = true
            }
        }
    }

    @IBAction private func addRemoveGroup(_ sender: NSSegmentedControl) {
        guard let cell = sender.cell as? NSSegmentedCell else {
            return
        }

        if NSColorPanel.sharedColorPanelExists {
            NSColorPanel.shared.close()
        }

        var row: Int
        switch SegmentTag(rawValue: cell.tag(forSegment: sender.selectedSegment)) {
        case .add:
            fTableView.beginUpdates()

            GroupsController.groups.addNewGroup()

            row = fTableView.numberOfRows

            fTableView.insertRows(at: IndexSet(integer: row), withAnimation: .slideUp)
            fTableView.endUpdates()

            fTableView.selectRowIndexes(IndexSet(integer: row), byExtendingSelection: false)
            fTableView.scrollRowToVisible(row)

            fSelectedColorNameField.window?.makeFirstResponder(fSelectedColorNameField)

        case .remove:
            row = fTableView.selectedRow

            fTableView.beginUpdates()

            GroupsController.groups.removeGroup(withRowIndex: row)

            fTableView.removeRows(at: IndexSet(integer: row), withAnimation: .slideUp)
            fTableView.endUpdates()

            if fTableView.numberOfRows > 0 {
                if row == fTableView.numberOfRows {
                    row -= 1
                }
                fTableView.selectRowIndexes(IndexSet(integer: row), byExtendingSelection: false)
                fTableView.scrollRowToVisible(row)
            }
        case .none:
            break
        }

        updateSelectedGroup()
    }

    @IBAction private func customDownloadLocationSheetShow(_ sender: Any?) {
        guard let window = fCustomLocationPopUp.window else {
            return
        }
        let panel = NSOpenPanel()

        panel.prompt = NSLocalizedString("Select", comment: "Preferences -> Open panel prompt")
        panel.allowsMultipleSelection = false
        panel.canChooseFiles = false
        panel.canChooseDirectories = true
        panel.canCreateDirectories = true

        panel.beginSheetModal(for: window) { result in
            let index = GroupsController.groups.index(forRow: self.fTableView.selectedRow)
            if result == .OK {
                let path = panel.urls[0].path
                GroupsController.groups.setCustomDownloadLocation(path, for: index)
                GroupsController.groups.setUsesCustomDownloadLocation(true, for: index)
            } else {
                if GroupsController.groups.customDownloadLocation(for: index) == nil {
                    GroupsController.groups.setUsesCustomDownloadLocation(false, for: index)
                }
            }
            self.refreshCustomLocationWithSingleGroup()
            self.fCustomLocationPopUp.selectItem(at: 0)
        }
    }

    @IBAction private func toggleUseCustomDownloadLocation(_ sender: Any) {
        let index = GroupsController.groups.index(forRow: fTableView.selectedRow)
        if fCustomLocationEnableCheck.state == .on {
            if GroupsController.groups.customDownloadLocation(for: index) != nil {
                GroupsController.groups.setUsesCustomDownloadLocation(true, for: index)
            } else {
                customDownloadLocationSheetShow(nil)
            }
        } else {
            GroupsController.groups.setUsesCustomDownloadLocation(false, for: index)
        }
        fCustomLocationPopUp.isEnabled = fCustomLocationEnableCheck.state == .on
    }

// MARK: - Rule editor

    @IBAction private func toggleUseAutoAssignRules(_ sender: Any) {
        let index = GroupsController.groups.index(forRow: fTableView.selectedRow)
        if fAutoAssignRulesEnableCheck.state == .on {
            if GroupsController.groups.autoAssignRules(for: index) != nil {
                GroupsController.groups.setUsesAutoAssignRules(true, for: index)
            } else {
                orderFrontRulesSheet(nil)
            }
        } else {
            GroupsController.groups.setUsesAutoAssignRules(false, for: index)
        }
        fAutoAssignRulesEditButton.isEnabled = fAutoAssignRulesEnableCheck.state == .on
    }

    @IBAction private func orderFrontRulesSheet(_ sender: Any?) {
        if groupRulesSheetWindow == nil {
            Bundle.main.loadNibNamed("GroupRules", owner: self, topLevelObjects: nil)
        }

        let index = GroupsController.groups.index(forRow: fTableView.selectedRow)
        let predicate = GroupsController.groups.autoAssignRules(for: index)
        ruleEditor.objectValue = predicate

        if ruleEditor.numberOfRows == 0 {
            ruleEditor.addRow(nil)
        }

        fTableView.window?.beginSheet(groupRulesSheetWindow, completionHandler: nil)
    }

    @IBAction private func cancelRules(_ sender: Any) {
        fTableView.window?.endSheet(groupRulesSheetWindow)

        let index = GroupsController.groups.index(forRow: fTableView.selectedRow)
        if GroupsController.groups.autoAssignRules(for: index) == nil {
            GroupsController.groups.setUsesAutoAssignRules(false, for: index)
            fAutoAssignRulesEnableCheck.state = .off
            fAutoAssignRulesEditButton.isEnabled = false
        }
    }

    @IBAction private func saveRules(_ sender: Any) {
        fTableView.window?.endSheet(groupRulesSheetWindow)

        let index = GroupsController.groups.index(forRow: fTableView.selectedRow)
        GroupsController.groups.setUsesAutoAssignRules(true, for: index)

        let predicate = ruleEditor.objectValue as? NSPredicate
        GroupsController.groups.setAutoAssignRules(predicate, for: index)

        fAutoAssignRulesEnableCheck.state = GroupsController.groups.usesAutoAssignRules(for: index) ? .on : .off
        fAutoAssignRulesEditButton.isEnabled = fAutoAssignRulesEnableCheck.state == .on
    }

    @objc
    func ruleEditorRowsDidChange(_ notification: Notification) {
        guard let ruleEditorScrollView = ruleEditor.enclosingScrollView else {
            return
        }

        let rowHeight = ruleEditor.rowHeight
        let bordersHeight = ruleEditorScrollView.frame.size.height - ruleEditorScrollView.contentSize.height

        let requiredRowCount = ruleEditor.numberOfRows
        let maxVisibleRowCount = Int((ruleEditor.window?.screen?.visibleFrame.height ?? 0) * 2 / 3 / rowHeight)

        ruleEditorHeightConstraint.constant = CGFloat(min(requiredRowCount, maxVisibleRowCount)) * rowHeight + bordersHeight
        ruleEditorScrollView.hasVerticalScroller = requiredRowCount > maxVisibleRowCount
    }

// MARK: - Private

    private func updateSelectedGroup() {
        fAddRemoveControl.setEnabled(fTableView.numberOfSelectedRows > 0, forSegment: SegmentTag.remove.rawValue)
        if fTableView.numberOfSelectedRows == 1 {
            let index = GroupsController.groups.index(forRow: fTableView.selectedRow)
            fSelectedColorView.color = GroupsController.groups.color(for: index)
            fSelectedColorView.isEnabled = true
            fSelectedColorNameField.stringValue = GroupsController.groups.name(for: index)
            fSelectedColorNameField.isEnabled = true

            refreshCustomLocationWithSingleGroup()

            fAutoAssignRulesEnableCheck.state = GroupsController.groups.usesAutoAssignRules(for: index) ? .on : .off
            fAutoAssignRulesEnableCheck.isEnabled = true
            fAutoAssignRulesEditButton.isEnabled = fAutoAssignRulesEnableCheck.state == .on
        } else {
            fSelectedColorView.color = .white
            fSelectedColorView.isEnabled = false
            fSelectedColorNameField.stringValue = ""
            fSelectedColorNameField.isEnabled = false
            fCustomLocationEnableCheck.isEnabled = false
            fCustomLocationPopUp.isEnabled = false
            fAutoAssignRulesEnableCheck.isEnabled = false
            fAutoAssignRulesEditButton.isEnabled = false
        }
    }

    private func refreshCustomLocationWithSingleGroup() {
        let index = GroupsController.groups.index(forRow: fTableView.selectedRow)

        let hasCustomLocation = GroupsController.groups.usesCustomDownloadLocation(for: index)
        fCustomLocationEnableCheck.state = hasCustomLocation ? .on : .off
        fCustomLocationEnableCheck.isEnabled = true
        fCustomLocationPopUp.isEnabled = hasCustomLocation

        if let location = GroupsController.groups.customDownloadLocation(for: index) {
            let pathTransformer = ExpandedPathToPathTransformer()
            fCustomLocationPopUp.item(at: 0)?.title = pathTransformer.transformedValue(location) as? String ?? ""
            let iconTransformer = ExpandedPathToIconTransformer()
            fCustomLocationPopUp.item(at: 0)?.image = iconTransformer.transformedValue(location) as? NSImage
        } else {
            fCustomLocationPopUp.item(at: 0)?.title = ""
            fCustomLocationPopUp.item(at: 0)?.image = nil
        }
    }
}

extension GroupsPrefsController: NSTableViewDataSource {
    func numberOfRows(in tableview: NSTableView) -> Int {
        return GroupsController.groups.numberOfGroups
    }
    func tableView(_ tableView: NSTableView, objectValueFor tableColumn: NSTableColumn?, row: Int) -> Any? {
        let groupsController = GroupsController.groups
        let groupsIndex = groupsController?.index(forRow: row) ?? 0

        let identifier = tableColumn?.identifier
        if identifier == NSUserInterfaceItemIdentifier("Color") {
            return groupsController?.image(for: groupsIndex)
        } else {
            return groupsController?.name(for: groupsIndex)
        }
    }

    func tableView(_ tableView: NSTableView, writeRowsWith rowIndexes: IndexSet, to pboard: NSPasteboard) -> Bool {
        pboard.declareTypes([ Self.kGroupTableViewDataType ], owner: self)
        pboard.setData(try? NSKeyedArchiver.archivedData(withRootObject: rowIndexes, requiringSecureCoding: true),
                       forType: Self.kGroupTableViewDataType)
        return true
    }

    func tableView(_ tableView: NSTableView, validateDrop info: NSDraggingInfo, proposedRow row: Int, proposedDropOperation dropOperation: NSTableView.DropOperation) -> NSDragOperation {
        let pasteboard = info.draggingPasteboard
        if pasteboard.types?.contains(Self.kGroupTableViewDataType) == true {
            fTableView.setDropRow(row, dropOperation: .above)
            return .generic
        }
        return []
    }

    func tableView(_ tableView: NSTableView, acceptDrop info: NSDraggingInfo, row: Int, dropOperation: NSTableView.DropOperation) -> Bool {
        let pasteboard = info.draggingPasteboard
        if pasteboard.types?.contains(Self.kGroupTableViewDataType) == true {
            guard let data = pasteboard.data(forType: Self.kGroupTableViewDataType) else {
                return true
            }
            guard let indexes = try? NSKeyedUnarchiver.unarchivedObject(ofClass: NSIndexSet.self, from: data) else {
                return true
            }

            let oldRow = indexes.firstIndex
            let newRow = oldRow < row ? row - 1 : row

            fTableView.beginUpdates()

            GroupsController.groups.moveGroup(atRow: oldRow, toRow: newRow)

            fTableView.moveRow(at: oldRow, to: newRow)
            fTableView.endUpdates()
        }
        return true
    }
}

extension GroupsPrefsController: NSTableViewDelegate {
    func tableViewSelectionDidChange(_ notification: Notification) {
        updateSelectedGroup()
    }
}

extension GroupsPrefsController: NSControlTextEditingDelegate {
    func controlTextDidEndEditing(_ obj: Notification) {
        if obj.object as? NSObject == fSelectedColorNameField {
            let index = GroupsController.groups.index(forRow: fTableView.selectedRow)
            GroupsController.groups.setName(fSelectedColorNameField.stringValue, for: index)
            fTableView.needsDisplay = true
        }
    }
}
